lib/semantics/resolve-names.cc

```diff
@@ -1539,121 +1643,81 @@ Symbol *SubprogramVisitor::GetSpecificFromGeneric(const parser::Name &name) {
-  EndDeclTypeSpec();
-  // TODO: TypeParamDefStmt
-}
-
-bool ResolveNamesVisitor::Pre(const parser::CommonBlockObject &x) {
-  BeginArraySpec();
-  return true;
-}
-void ResolveNamesVisitor::Post(const parser::CommonBlockObject &x) {
-  ClearArraySpec();
-  // TODO: CommonBlockObject
-}
-
-void ResolveNamesVisitor::Post(const parser::ComponentDecl &) {
-  ClearArraySpec();
-}
-
-bool ResolveNamesVisitor::Pre(const parser::PrefixSpec &x) {
-  return true;  // TODO
-}
-
-bool ResolveNamesVisitor::CheckUseError(
-    const SourceName &name, const Symbol &symbol) {
-  const auto *details = symbol.detailsIf<UseErrorDetails>();
-  if (!details) {
-    return false;
-  }
-  Message &msg{Say(name, "Reference to '%s' is ambiguous"_err_en_US)};
-  for (const auto &pair : details->occurrences()) {
-    const SourceName &location{*pair.first};
-    const SourceName &moduleName{pair.second->name()};
-    msg.Attach(location,
-        MessageFormattedText{"'%s' was use-associated from module '%s'"_en_US,
-            name.ToString().data(), moduleName.ToString().data()});
-  }
-  return true;
+void DeclarationVisitor::Post(const parser::DimensionStmt::Declaration &x) {
+  const auto &name = std::get<parser::Name>(x.t);
+  DeclareObjectEntity(name, Attrs{});
 }
 
-void ResolveNamesVisitor::Post(const parser::ProcedureDesignator &x) {
-  if (const auto *name = std::get_if<parser::Name>(&x.u)) {
-    Symbol &symbol{MakeSymbol(name->source)};
-    if (symbol.has<UnknownDetails>()) {
-      if (isImplicitNoneExternal() && !symbol.attrs().test(Attr::EXTERNAL)) {
-        Say(*name,
-            "'%s' is an external procedure without the EXTERNAL"
-            " attribute in a scope with IMPLICIT NONE(EXTERNAL)"_err_en_US);
-      }
-      symbol.attrs().set(Attr::EXTERNAL);
-      symbol.set_details(SubprogramDetails{});
-    } else if (CheckUseError(name->source, symbol)) {
-      // error was reported
-    } else if (!symbol.isSubprogram()) {
-      auto *details = symbol.detailsIf<EntityDetails>();
-      if (!details || !details->isArray()) {
-        Say(*name,
-            "Use of '%s' as a procedure conflicts with its declaration"_err_en_US)
-            .Attach(symbol.name(),
-                MessageFormattedText{"Declaration of '%s'"_en_US,
-                    symbol.name().ToString().data()});
-      }
+void DeclarationVisitor::Post(const parser::EntityDecl &x) {
+  // TODO: may be under StructureStmt
+  const auto &name{std::get<parser::ObjectName>(x.t)};
+  // TODO: CoarraySpec, CharLength, Initialization
+  Attrs attrs{attrs_ ? *attrs_ : Attrs{}};
+  if (!arraySpec().empty()) {
+    DeclareObjectEntity(name, attrs);
+  } else {
+    Symbol &symbol{DeclareEntity<EntityDetails>(name, attrs)};
+    if (auto &type = GetDeclTypeSpec()) {
+      SetType(name.source, symbol, *type);
     }
   }
 }
 
-bool ResolveNamesVisitor::Pre(const parser::AsynchronousStmt &x) {
+bool DeclarationVisitor::Pre(const parser::AsynchronousStmt &x) {
```

Comment:
Be advised, ASYNCHRONOUS and VOLATILE have their own peculiar behavior in BLOCK constructs and contained subprograms -- when they are the only declaration statement used on a name that is already available through host association (or whatever the right term would be in the case of BLOCK), they don't introduce a new symbol, but just give it a scoped attribute.

Reply:
HandleAttributeStmt() has special handling for ASYNCHRONOUS and VOLATILE for symbols that are use-associated. It will need the same when host association is implemented too.